src/stackmap_checker/guard.c

```diff
@@ -0,0 +1,350 @@
+#define UNW_LOCAL_ONLY
+
+#include <stdint.h>
+#include <unistd.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+#include <err.h>
+#include <signal.h>
+#include <libunwind.h>
+#include "stmap.h"
+#include "utils.h"
+
+#define MAX_CALL_STACK_DEPTH 256
+
+// XXX These currently need to be global, since they need to be visible to
+// jump.s
+// The label to jump to whenever a guard fails.
+extern void restore_and_jmp(void);
+// The address to jump to.
+uint64_t addr = 0;
+
+// The state of the call stack.
+typedef struct CallStackState {
+    // The address of the return address of each frame.
+    uint64_t *ret_addrs;
+    // The base pointers.
+    unw_word_t *bps;
+    // The 16 registers recorded for each frame.
+    unw_word_t **registers;
+    // The stack map records which correspond to each call on the stack.
+    stack_map_record_t *records;
+    // The number of frames on the stack.
+    uint32_t depth;
+} call_stack_state_t;
+
+/*
+ * Return the state of the call stack.
+ */
+call_stack_state_t* get_call_stack_state(unw_cursor_t cursor,
+                                         unw_context_t context)
+{
+    uint64_t *ret_addrs = calloc(MAX_CALL_STACK_DEPTH, sizeof(uint64_t));
+    unw_word_t *bps = calloc(MAX_CALL_STACK_DEPTH, sizeof(unw_word_t));
+    unw_word_t **registers = calloc(MAX_CALL_STACK_DEPTH, sizeof(unw_word_t *));
+    uint32_t frame = 0;
+    while (unw_step(&cursor) > 0) {
+        unw_word_t off, pc;
+        unw_get_reg(&cursor, UNW_REG_IP, &pc);
+        if (!pc) {
+            break;
+        }
+
+        char fun_name[128];
+        unw_get_proc_name(&cursor, fun_name, sizeof(fun_name), &off);
+
+        // Stop when main is reached.
+        if (!strcmp(fun_name, "main")) {
+            break;
+        }
+
+        if (!registers[frame]) {
+            // 16 registers are saved for each frame
+            registers[frame] = calloc(16, sizeof(unw_word_t));
```

Comment:
Have you used valgrind on this code to check there are no memory errors?

Reply:
Yes, I have.

Reply:
Good job!